File: src/main/java/nu/nldv/runapp/util/GpxMapper.java
package nu.nldv.runapp.util;

import nu.nldv.runapp.model.Coordinates;
import nu.nldv.runapp.model.Point;
import nu.nldv.runapp.model.Segment;
import nu.nldv.runapp.model.Track;
import nu.nldv.runapp.model.gpx.GpxType;
import nu.nldv.runapp.model.gpx.TrkType;
import nu.nldv.runapp.model.gpx.TrksegType;
import nu.nldv.runapp.model.gpx.WptType;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

/**
 * Created with IntelliJ IDEA.
 * User: nicklas
 * Date: 2014-01-01
 * Time: 16:53
 */
@Service
public class GpxMapper {

    @Autowired
    private TimeCalculator timeCalculator;

    public List<Track> map(GpxType gpx) {
        List<Track> tracks = new ArrayList<>();
        for (TrkType trk: gpx.getTrk()) {
            Track track = new Track();
            track.setSegments(mapSegments(trk.getTrkseg()));
            track.setStartDate(track.getSegments().get(0).getPoints().get(0).getDate());
            track.setDuration(timeCalculator.getDurationBetweenDates(track.getStartDate(),getLastPointDate(track)));
            tracks.add(track);
        }

        return tracks;
    }

    private Date getLastPointDate(Track track) {
        Segment lastSegment= getLastSegment(track);
        return lastSegment.getPoints().get(lastSegment.getPoints().size()-1).getDate();

    }

    private Segment getLastSegment(Track track) {
        return track.getSegments().get(track.getSegments().size()-1);
    }

    private List<Segment> mapSegments(List<TrksegType> trksegs) {
        List<Segment> segments = new ArrayList<>();
        for (TrksegType trksegType : trksegs) {
            Segment segment = new Segment();
            segment.setPoints(mapPoints(trksegType.getTrkpt()));
            segments.add(segment);
        }

        return segments;
    }

    private List<Point> mapPoints(List<WptType> trkpts) {
        List<Point> points = new ArrayList<>();
        for (WptType trkpt : trkpts) {
            Point point = new Point();
            point.setElevation(trkpt.getEle().doubleValue());
            point.setCoordinates(mapCoordinates(trkpt));
            point.setDate(trkpt.getTime().toGregorianCalendar().getTime());
            points.add(point);
        }

        return points;
    }

    private Coordinates mapCoordinates(WptType trkpt) {
        Coordinates coordinate=new Coordinates();
        coordinate.setLatitude(trkpt.getLat().doubleValue());
        coordinate.setLongitude(trkpt.getLon().doubleValue());
        return coordinate;
    }
}

File: app/util/GpxMapper.scala
package util

import nu.nldv.parsethatgpx.model.{Point, Segment, Gpx}
import model.{Coordinates, Track}
import org.joda.time.Period

/**
 * Created with IntelliJ IDEA.
 * User: nicklas
 * Date: 2014-03-23
 * Time: 20:16
 */
object GpxMapper {

  def duration(segments: List[Segment]): Option[String] = {
    val duration = Period.millis((segments.takeRight(1).head.points.takeRight(1).head.time.getMillis - segments.head.points.head.time.getMillis).toInt).toStandardDuration
    val allSeconds = duration.getStandardSeconds
    val hours = allSeconds / 3600
    val minutes = (allSeconds - hours * 3600) / 60
    val seconds = (allSeconds - (hours * 3600) - (minutes * 60))
    val str = if (hours > 10) "" + hours else "0" + hours + ":" + (if (minutes > 10) minutes else "0" + minutes) + ":" + (if (seconds > 10) seconds else "0" + seconds)
    Some(str)
  }

  def points(points: List[Point]): List[model.Point] = points.map(point => model.Point(point.elevation, Coordinates(point.latitude, point.longitude), point.time, None))

  def toSegments(segments: List[Segment]): List[model.Segment] = segments.map(segment => model.Segment(points(segment.points)))

  def toTrack(gpx: Gpx): Track = Track(id = -1,
    segments = toSegments(gpx.track.segments),
    duration = duration(gpx.track.segments),
    distance = None,
    pace = None,
    date = gpx.track.time,
    startDate = None)

}
